tests/integration/linux_bridge_test.py

```diff
@@ -1282,3 +1282,32 @@ def test_controller_detach_from_linux_bridge(bridge0_with_port0):
         )
         == 0
     )
+
+
+def test_port_remove_and_validate_port_controller(bond0):
+    bond0[Interface.KEY][0][Interface.CONTROLLER] = TEST_BRIDGE0
+    bond_name = bond0[Interface.KEY][0][Interface.NAME]
+    desired_state = bond0
+    bridge_config_state = _create_bridge_subtree_config(())
+    with linux_bridge(TEST_BRIDGE0, bridge_config_state) as bridge0:
+        desired_state = bond0
+        bridge_state = bridge0[Interface.KEY][0]
+        desired_state[Interface.KEY].append(bridge_state)
+        libnmstate.apply(desired_state)
+        assertlib.assert_state_match(desired_state)
+
+        libnmstate.apply(
+            {
+                Interface.KEY: [
+                    {
+                        Interface.NAME: bond_name,
+                        Interface.STATE: InterfaceState.ABSENT,
+                    }
+                ]
+            }
+        )
+        assertlib.assert_absent(bond_name)
+
+        libnmstate.apply(desired_state)
```

Comment:
The key reproducer of original problem is bond0 has controller set to br0, but br0 already exists with empty ports.

It is very hard to summarize that from your test code.

How about:


```python

@pytest.fixture
def empty_bridge():
    # create a empty bridge then yiled and clean up afterwards

def test_attach_bond_to_empty_bridge(empty_bridge, bond0):
    # apply a desire state with bond0 holding controller property and br0 mentioned with name and type only.
    # no need to clean up and assert cleanup as they are not the focus of this test.
    # the fixture should clean up everything. 

```

Reply:
https://issues.redhat.com/browse/RHEL-31977, 
applying the bond0 holding controller property and empty bridge can not reproduce the issue, we have to delete the bond interface in the second step. Then in the final step,  applying the bond0 holding controller property and empty bridge again can reproduce the issue.

Reply:
The test case should reflect the key reproducer not unrelated environment setup.

Reply:
Thanks for reviewing it, I updated the patch again.